### PR TITLE
EOS-13985:Implement nfs_req.py and hist.py <utils repo>

### PR DESCRIPTION
**Pull request by pratyush-seagate**
# EOS-13985:Implement nfs_req.py and hist.py <cortxfs repo>

## Checklist
- [x] **Compilation:** _This patch does not break compilation_
- [x] **Merge conflicts:** _This patch has been squashed and re-based, it can be merged using fast-forward merge_
- [] **Code review:** _All discussions have been resolved_
- [x] **Sanity Testing:** _All Unit tests are passing and able to do mount and io operations works from NFS client_
- [x] **Documentation:** _This patch and merge request have up to date description_
- [x] **Unit Testing and debugging:** _Both single and multi node changes are unit tested_

## Commit Message
            commit 65c25247aef3e380d092441a87647894ebdf773e
            Author: pratyush-seagate <pratyush.k.khan@seagate.com>
            Date:   Mon Oct 12 17:23:50 2020 -0600

            EOS-13985:Implement nfs_req.py and hist.py <Part 1, utils repo>
            List of modified files:
                modified:   c-utils/src/addb/plugin.c
                modified:   c-utils/src/include/perf/perf-counters.h
                modified:   c-utils/src/test/perf/perf-manual-test.c
            Change description:
                This repo's change is to track the operation id of immediate callers in map performance counters which will be used by addb2db_cortxfs.py. Thi
            Unit test:
                    Run NFS READ IO test, generate DB from addb2db_cortxfs.py, use that DB to generate a png file which shows call graph constructed from that
            TODO: Implement timeline and histogram tools
            The generate graph image along with DB file and raw dump files are saved at: /home/751521/JIRA/EOS-13985
                    [root@ssc-vm-c-0040 /]# python3 /root/cortxfs_req.py -d ./cortxfs_perfc_demo_decoded_1.db 0x537B -o graph
                    Creating cortxfs performance call graphs for fsal_op_id 21371, from db file ./cortxfs_perfc_demo_decoded_1.db, o/p graph png file graph
                    Graph render completed

### CODE REVIEW

File: c-utils/src/include/perf/perf-counters.h [hunk 1]
```diff
@@ -376,6 +376,26 @@ static inline struct perfc_tls_ctx perfc_tls_get_top(void)
...
+	if (perfstack == NULL) {
+	    return ptctx;
+	}
+
+	if (perfstack->top == -1) {
+	    return ptctx;
+	}
+
+	if (perfstack->top == 0) {
+		return perfstack->pstack[perfstack->top];
```
Comment (AtitaShirwaikar) [inline c-utils/src/include/perf/perf-counters.h:393]: extra tab

**Review by vimalraghwani20 — APPROVED** [review]
Looks good to me

File: c-utils/src/include/perf/perf-counters.h [hunk 2]
```diff
@@ -376,6 +376,26 @@ static inline struct perfc_tls_ctx perfc_tls_get_top(void)
 	return perfstack->pstack[perfstack->top];
 }
 
+static inline struct perfc_tls_ctx perfc_tls_get_prev(void)
+{
+	struct perfc_tls_ctx ptctx = {.opid = PERFC_INVALID_ID};
+	struct perfc_callstack *perfstack = pthread_getspecific(perfc_tls_key);
+
+	if (perfstack == NULL) {
```
Comment (AtitaShirwaikar) [inline c-utils/src/include/perf/perf-counters.h:384]: This and the condition below can be combined into a single check

File: c-utils/src/addb/plugin.c [hunk 3]
```diff
@@ -183,7 +183,8 @@ static struct m0_addb2__id_intrp gs_curr_ids[] = {
             &decode_perfc_function_tags,
             &decode_perfc_entry_type,
             &hex, // operation id
-            &hex // mapping with opertation id
+            &hex, // mapping with origin opertation id
```
Comment (AtitaShirwaikar) [inline c-utils/src/addb/plugin.c:186]: Nit pick-Operation spelling

**Review by AtitaShirwaikar — APPROVED** [review]
Minor comment

File: c-utils/src/include/perf/perf-counters.h [hunk 4]
```diff
@@ -285,11 +285,11 @@ enum perfc_entity_maps {
  *	- related_opid - caller generated tag for related operation id
  *	- __VA_ARGS_ - attribute-specific arguments.
  */
-#define PERFC_MAP_V2(__mod, __fn_tag, map_tag, opid, related_opid, ...) do {\
+#define PERFC_MAP_V2(__mod, __fn_tag, map_tag, opid, related_opid, caller_opid,...) do {\
```
Comment (AtitaShirwaikar) [inline c-utils/src/include/perf/perf-counters.h:288]: Update the where arguments description above for this new added caller_opid parameter

File: c-utils/src/include/perf/perf-counters.h [hunk 5]
```diff
@@ -394,10 +414,12 @@ static inline struct perfc_tls_ctx perfc_tls_get_top(void)
 
 #define perfc_trace_map(fn_tag, map_tag, _opid, ...) do {		\
 	struct perfc_tls_ctx ptctx = perfc_tls_get_origin();		\
```
Comment (AtitaShirwaikar) [inline c-utils/src/include/perf/perf-counters.h:416]: Would be good if  ptcctx  can be renamed.The name is confusing.Maybe ptpctx, to indicate previous context

Comment (pratyush-seagate) [reply c-utils/src/include/perf/perf-counters.h:416]: ptctx's <ptc> prefix is abbrev for "perfc_tls_ctx". Also this is a local variable, so I think we can leave it as it is.

**Review by AtitaShirwaikar — APPROVED** [review]
+1

**Review by shahidSTX — APPROVED** [review]
LGTM.

**Review by sachinpunadikar — APPROVED** [review]
+1

**Pull request merged**